### ISSUE TITLE
Breakpoint doesn't work after restart

### ISSUE THREAD

**Issue by karliss** [radareorg/radare2#17212]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Arch Linux x86_64
| File format of the file you reverse (mandatory)      | ELF
| Architecture/bits of the file (mandatory)            | x86_64
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.5.0-git 26401 @ linux-x86-64 git.4.4.0-378-gd6c9bd454 commit: d6c9bd45429287220ffe16cc644b48164ff8a463 build: 2020-07-05__11:03:14

### Expected behavior

After restarting debug session breakpoints in the executable are preserved

### Actual behavior

`pd` and `db` indicates that breakpoint is valid and in the correct position but `dc` runs the executable without hitting the breakpoint.

### Steps to reproduce the behavior 
- `r2 /usr/bin/ls`
- `db main`
- `doo`
- `dc` until you hit the breakpoint
- `doo`
- `db`, `pd 10 @ main`
- `dc` until executable finishes, observe that breakpoint isn't hit

```
[karlis@desktop1 radare2]$ r2 /usr/bin/ls
Dynamic tag 12 not handled
Dynamic tag 13 not handled
Dynamic tag 25 not handled
Dynamic tag 27 not handled
Dynamic tag 26 not handled
Dynamic tag 28 not handled
Dynamic tag 1879047925 not handled
Dynamic tag 21 not handled
 -- Add comments using the ';' key in visual mode or the 'CC' command from the radare2 shell
[0x00005b20]> db main
[0x00005b20]> doo
Process with PID 33946 started...
= attach 33946 33946
File dbg:///usr/bin/ls  reopened in read-write mode
Dynamic tag 12 not handled
Dynamic tag 13 not handled
Dynamic tag 25 not handled
Dynamic tag 27 not handled
Dynamic tag 26 not handled
Dynamic tag 28 not handled
Dynamic tag 1879047925 not handled
Dynamic tag 21 not handled
33946
[0x7fb498f20100]> dc
hit breakpoint at: 55f96fc610a0
[0x55f96fc610a0]> pd 8 @ main
            ;-- main:
            ;-- rax:
            ;-- rip:
            0x55f96fc610a0 b    4157           push r15
            0x55f96fc610a2      4156           push r14
            0x55f96fc610a4      4155           push r13
            0x55f96fc610a6      4154           push r12
            0x55f96fc610a8      4189fc         mov r12d, edi
            0x55f96fc610ab      55             push rbp
            0x55f96fc610ac      4889f5         mov rbp, rsi
            0x55f96fc610af      53             push rbx
[0x55f96fc610a0]> doo
child received signal 9
Stepping failed!
Process with PID 33957 started...
= attach 33957 33957
File dbg:///usr/bin/ls  reopened in read-write mode
Dynamic tag 12 not handled
Dynamic tag 13 not handled
Dynamic tag 25 not handled
Dynamic tag 27 not handled
Dynamic tag 26 not handled
Dynamic tag 28 not handled
Dynamic tag 1879047925 not handled
Dynamic tag 21 not handled
33957
[0x7f3fd557c100]> db
0x5629e7a2d0a0 - 0x5629e7a2d0a1 1 --x sw break enabled valid cmd="" cond="" name="main" module=""
[0x7f3fd557c100]> pd 8 @ main
            ;-- main:
            0x5629e7a2d0a0 b    4157           push r15
            0x5629e7a2d0a2      4156           push r14
            0x5629e7a2d0a4      4155           push r13
            0x5629e7a2d0a6      4154           push r12
            0x5629e7a2d0a8      4189fc         mov r12d, edi
            0x5629e7a2d0ab      55             push rbp
            0x5629e7a2d0ac      4889f5         mov rbp, rsi
            0x5629e7a2d0af      53             push rbx
[0x7f3fd557c100]> dc
AUTHORS.md  configure	       config-user.mk	   COPYING.LESSER  Dockerfile.travis  libr	   meson_options.txt  plugins.android.cfg  plugins.emscripten.cfg  plugins.static.cfg	     README.md	     test		    Vagrantfile.sh
autogen.sh  configure.acr      config-user.mk.acr  DEVELOPERS.md   Doxyfile	      Makefile	   mk		      plugins.bin.cfg	   plugins.ios.cfg	   plugins.static.nogpl.cfg  shlr	     travis-extract-var.sh
binr	    configure.hook     CONTRIBUTING.md	   doc		   env.sh	      man	   package.json       plugins.cfg	   plugins.ios-store.cfg   plugins.tiny.cfg	     snapcraft.yaml  travis-script
CODEOWNERS  configure-plugins  COPYING		   Dockerfile	   global.mk	      meson.build  pkgcfg	      plugins.def.cfg	   plugins.nogpl.cfg	   radare2.spec		     sys	     Vagrantfile
[0x7f3fd540b0f1]> dc
child exited with status 0

==> Process finished
```

### Additional Logs, screenshots, source-code,  configuration dump, ...
[ls.zip](https://github.com/radareorg/radare2/files/4874705/ls.zip)

**Comment by trufae** [comment]
Breakpoints are set by offset not by name. So if you restart the system is probably using aslr to place the bin somewhere else and it fails to set the bp in the main because the main is somewhere else.

Theres an issue to support named and future breakpoints

**Comment by karliss** [comment]
@trufae When starting/stopping debugging breakpoints in the main executable are relocated the same way as flags. That's how it has worked for a while and why it is possible to set breakpoint before starting to debug. In the log output above you can also see that I verified using `db` and `pd` that after restarting breakpoint was supposedly relocated to correct address.

**Comment by karliss** [comment]
Described steps work in 4.2.1.

**Comment by karliss** [comment]
Additional observations:
If the executable in the first debug session finishes and I reopen it in non-debug mode before reopening in debug mode then it works.
db -> doo -> dc (and hit bp) -> dc x2 until executable exits -> oo -> doo ->dc (and hit bp) works
db -> doo -> dc (and hit bp) -> doo -> dc  doesn't work

**Comment by karliss** [comment]
I started debugging this and it seems that some parts somewhat work due to being double wrong.
```
	// This means we did a first run, so we probably attached to all possible threads already.
	// So check if the requested thread is being traced already. If yes: skip
	if (dbg->threads && !r_list_find (dbg->threads, &pid, &match_pid)) {
		goto out;
	}
	linux_attach_single_pid (dbg, pid);
```
From the comment it is supposed to skip when pid is already in the list but the if does opposite. `match_pid` returns a boolean int `pid == th->pid` 0 = doesn't match, but r_list_find expects a comparison method which returns -1, 0, 1 with 0 meaning match.

**Comment by abcSup** [comment]
> I started debugging this and it seems that some parts somewhat work due to being double wrong.
> 
> ```
> 	// This means we did a first run, so we probably attached to all possible threads already.
> 	// So check if the requested thread is being traced already. If yes: skip
> 	if (dbg->threads && !r_list_find (dbg->threads, &pid, &match_pid)) {
> 		goto out;
> 	}
> 	linux_attach_single_pid (dbg, pid);
> ```
> 
> From the comment it is supposed to skip when pid is already in the list but the if does opposite. `match_pid` returns a boolean int `pid == th->pid` 0 = doesn't match, but r_list_find expects a comparison method which returns -1, 0, 1 with 0 meaning match.

I encounter this problem while working on another PR that fixes an error when attaching a multithreaed process. In addition, `match_pid` is also wrong, especially it tries to treat the `iter->data` as `RDebug` instead of `RDebugPid`.

**Comment by karliss** [comment]
As for the actual problem so far it seems that dbg->reason state isn't properly cleaned up or reinitialized during `doo->doo` transition and the old `dbg->reason` information prevents breakpoints from being properly restored as the debugger thinks that it is still handling recoil for the breakpoints in first process.
